2026-07-06T19:26:50.3853082Z ##[group]Run RUSTDOCFLAGS="-D warnings" cargo doc --workspace --no-deps
2026-07-06T19:26:50.3854031Z [36;1mRUSTDOCFLAGS="-D warnings" cargo doc --workspace --no-deps[0m
2026-07-06T19:26:50.3888101Z shell: /usr/bin/bash -e {0}
2026-07-06T19:26:50.3888364Z env:
2026-07-06T19:26:50.3888568Z   CARGO_TERM_COLOR: always
2026-07-06T19:26:50.3888813Z   CARGO_BUILD_JOBS: 2
2026-07-06T19:26:50.3889036Z   RUSTFLAGS: -D warnings
2026-07-06T19:26:50.3889286Z   CARGO_HOME: /home/runner/.cargo
2026-07-06T19:26:50.3889540Z   CARGO_INCREMENTAL: 0
2026-07-06T19:26:50.3889768Z ##[endgroup]
2026-07-06T19:26:50.5043218Z [1m[92m   Compiling[0m unicode-ident v1.0.24
2026-07-06T19:26:50.5044001Z [1m[92m    Checking[0m cfg-if v1.0.4
2026-07-06T19:26:50.5202946Z [1m[92m   Compiling[0m num-traits v0.2.19
2026-07-06T19:26:50.5450502Z [1m[92m   Compiling[0m proc-macro2 v1.0.106
2026-07-06T19:26:50.5749241Z [1m[92m    Checking[0m bitflags v2.13.0
2026-07-06T19:26:50.6728390Z [1m[92m    Checking[0m libm v0.2.16
2026-07-06T19:26:50.9111724Z [1m[92m   Compiling[0m quote v1.0.46
2026-07-06T19:26:51.0335474Z [1m[92m   Compiling[0m syn v2.0.118
2026-07-06T19:26:51.1688522Z [1m[92m    Checking[0m libloading v0.8.9
2026-07-06T19:26:51.2182412Z [1m[92m    Checking[0m equivalent v1.0.2
2026-07-06T19:26:51.2346146Z [1m[92m    Checking[0m log v0.4.33
2026-07-06T19:26:51.6236095Z [1m[92m    Checking[0m foldhash v0.2.0
2026-07-06T19:26:51.6679589Z [1m[92m    Checking[0m allocator-api2 v0.2.21
2026-07-06T19:26:51.8267985Z [1m[92m    Checking[0m hashbrown v0.16.1
2026-07-06T19:26:52.1385530Z [1m[92m    Checking[0m libc v0.2.186
2026-07-06T19:26:52.6355588Z [1m[92m    Checking[0m bit-vec v0.9.1
2026-07-06T19:26:52.7222053Z [1m[92m    Checking[0m hashbrown v0.17.1
2026-07-06T19:26:53.0269618Z [1m[92m    Checking[0m unicode-width v0.2.2
2026-07-06T19:26:53.1272531Z [1m[92m    Checking[0m once_cell v1.21.4
2026-07-06T19:26:53.1963501Z [1m[92m    Checking[0m codespan-reporting v0.13.1
2026-07-06T19:26:53.3448950Z [1m[92m    Checking[0m indexmap v2.14.0
2026-07-06T19:26:53.3889532Z [1m[92m   Compiling[0m thiserror-impl v2.0.18
2026-07-06T19:26:53.7011986Z [1m[92m   Compiling[0m zerocopy-derive v0.8.52
2026-07-06T19:26:54.0962909Z [1m[92m    Checking[0m thiserror v2.0.18
2026-07-06T19:26:54.1167766Z [1m[92m   Compiling[0m bytemuck_derive v1.10.2
2026-07-06T19:26:54.5896901Z [1m[92m    Checking[0m zerocopy v0.8.52
2026-07-06T19:26:54.8209821Z [1m[92m    Checking[0m bytemuck v1.25.0
2026-07-06T19:26:54.9335502Z [1m[92m    Checking[0m bit-set v0.9.1
2026-07-06T19:26:54.9876216Z [1m[92m    Checking[0m spirv v0.4.0+sdk-1.4.341.0
2026-07-06T19:26:55.6458987Z [1m[92m    Checking[0m rustc-hash v1.1.0
2026-07-06T19:26:55.8911877Z [1m[92m    Checking[0m smallvec v1.15.2
2026-07-06T19:26:56.1437119Z [1m[92m    Checking[0m scopeguard v1.2.0
2026-07-06T19:26:56.1644758Z [1m[92m    Checking[0m hexf-parse v0.2.1
2026-07-06T19:26:56.2007422Z [1m[92m    Checking[0m raw-window-handle v0.6.2
2026-07-06T19:26:56.2778641Z [1m[92m    Checking[0m foldhash v0.1.5
2026-07-06T19:26:56.3212631Z [1m[92m    Checking[0m arrayvec v0.7.7
2026-07-06T19:26:56.4018399Z [1m[92m    Checking[0m hashbrown v0.15.5
2026-07-06T19:26:56.6814400Z [1m[92m    Checking[0m wgpu-types v29.0.3
2026-07-06T19:26:56.9482193Z [1m[92m    Checking[0m half v2.7.1
2026-07-06T19:26:57.1167546Z [1m[92m    Checking[0m naga v29.0.3
2026-07-06T19:26:57.7981895Z [1m[92m    Checking[0m lock_api v0.4.14
2026-07-06T19:26:57.8923242Z [1m[92m    Checking[0m parking_lot_core v0.9.12
2026-07-06T19:26:57.9711561Z [1m[92m    Checking[0m ash v0.38.0+1.3.281
2026-07-06T19:27:01.6400658Z [1m[92m    Checking[0m dlib v0.5.3
2026-07-06T19:27:01.6645520Z [1m[92m    Checking[0m gpu-descriptor-types v0.2.0
2026-07-06T19:27:01.7100009Z [1m[92m    Checking[0m presser v0.3.1
2026-07-06T19:27:01.7424773Z [1m[92m   Compiling[0m serde_core v1.0.228
2026-07-06T19:27:01.8595449Z [1m[92m    Checking[0m gpu-descriptor v0.3.2
2026-07-06T19:27:01.9329217Z [1m[92m    Checking[0m wayland-sys v0.31.11
2026-07-06T19:27:01.9881138Z [1m[92m    Checking[0m wgpu-naga-bridge v29.0.3
2026-07-06T19:27:02.0160818Z [1m[92m    Checking[0m parking_lot v0.12.5
2026-07-06T19:27:02.0980925Z [1m[92m    Checking[0m crossbeam-utils v0.8.21
2026-07-06T19:27:02.2794789Z [1m[92m    Checking[0m khronos-egl v6.0.0
2026-07-06T19:27:02.4570762Z [1m[92m    Checking[0m ordered-float v5.3.0
2026-07-06T19:27:02.5747460Z [1m[92m    Checking[0m glow v0.17.0
2026-07-06T19:27:03.2305602Z [1m[92m    Checking[0m utf8parse v0.2.2
2026-07-06T19:27:03.2538239Z [1m[92m    Checking[0m profiling v1.0.18
2026-07-06T19:27:03.2683380Z [1m[92m    Checking[0m renderdoc-sys v1.1.0
2026-07-06T19:27:03.2972276Z [1m[92m    Checking[0m anstyle-parse v1.0.0
2026-07-06T19:27:03.3517774Z [1m[92m    Checking[0m crossbeam-epoch v0.9.18
2026-07-06T19:27:04.0370739Z [1m[92m    Checking[0m gpu-allocator v0.28.0
2026-07-06T19:27:04.1597693Z [1m[92m    Checking[0m wgpu-hal v29.0.3
2026-07-06T19:27:04.6061055Z [1m[92m   Compiling[0m serde v1.0.228
2026-07-06T19:27:04.7104778Z [1m[92m    Checking[0m is_terminal_polyfill v1.70.2
2026-07-06T19:27:04.7297003Z [1m[92m    Checking[0m anstyle v1.0.14
2026-07-06T19:27:04.8038224Z [1m[92m    Checking[0m colorchoice v1.0.5
2026-07-06T19:27:04.8256162Z [1m[92m    Checking[0m anstyle-query v1.1.5
2026-07-06T19:27:04.8475213Z [1m[92m    Checking[0m anstream v1.0.0
2026-07-06T19:27:04.9455569Z [1m[92m    Checking[0m crossbeam-deque v0.8.6
2026-07-06T19:27:05.0173815Z [1m[92m    Checking[0m zmij v1.0.21
2026-07-06T19:27:05.1382657Z [1m[92m   Compiling[0m serde_derive v1.0.228
2026-07-06T19:27:05.7241991Z [1m[92m    Checking[0m wgpu-core-deps-windows-linux-android v29.0.3
2026-07-06T19:27:05.7392412Z [1m[92m    Checking[0m strsim v0.11.1
2026-07-06T19:27:05.8023117Z [1m[92m    Checking[0m clap_lex v1.1.0
2026-07-06T19:27:05.8566257Z [1m[92m    Checking[0m itoa v1.0.18
2026-07-06T19:27:05.9081416Z [1m[92m    Checking[0m memchr v2.8.2
2026-07-06T19:27:06.1432801Z [1m[92m    Checking[0m serde_json v1.0.150
2026-07-06T19:27:06.5669181Z [1m[92m   Compiling[0m clap_derive v4.6.1
2026-07-06T19:27:07.2692704Z [1m[92m    Checking[0m clap_builder v4.6.0
2026-07-06T19:27:07.6072805Z [1m[92m    Checking[0m wgpu-core v29.0.3
2026-07-06T19:27:08.1521312Z [1m[92m    Checking[0m rayon-core v1.13.0
2026-07-06T19:27:08.3314617Z [1m[92m    Checking[0m static_assertions v1.1.0
2026-07-06T19:27:08.3479385Z [1m[92m    Checking[0m either v1.16.0
2026-07-06T19:27:08.4211736Z [1m[92m    Checking[0m rayon v1.12.0
2026-07-06T19:27:09.5199051Z [1m[92m    Checking[0m clap v4.6.1
2026-07-06T19:27:09.5373013Z [1m[92m    Checking[0m neat-core v0.2.5 (/home/runner/work/NEAT-AI-scorer/NEAT-AI-core/neat-core)
2026-07-06T19:27:10.0631165Z [1m[92m    Checking[0m pollster v0.4.0
2026-07-06T19:27:10.1384656Z [1m[92m    Checking[0m wgpu v29.0.3
2026-07-06T19:27:11.1948975Z [1m[92m Documenting[0m rust_scorer v1.0.6 (/home/runner/work/NEAT-AI-scorer/NEAT-AI-scorer/rust_scorer)
2026-07-06T19:27:11.1950935Z [1m[92m    Checking[0m rust_scorer v1.0.6 (/home/runner/work/NEAT-AI-scorer/NEAT-AI-scorer/rust_scorer)
2026-07-06T19:27:12.0156226Z [1m[91merror[0m[1m: unresolved link to `reset`[0m
2026-07-06T19:27:12.0156927Z   [1m[94m--> [0mrust_scorer/src/multi_score.rs:61:31
2026-07-06T19:27:12.0157400Z    [1m[94m|[0m
2026-07-06T19:27:12.0158078Z [1m[94m61[0m [1m[94m|[0m /// Usage from a test: call [`reset`], run one scoring invocation, then assert
2026-07-06T19:27:12.0159093Z    [1m[94m|[0m                               [1m[91m^^^^^[0m [1m[91mno item named `reset` in scope[0m
2026-07-06T19:27:12.0159690Z    [1m[94m|[0m
2026-07-06T19:27:12.0160314Z    [1m[94m= [0m[1mhelp[0m: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
2026-07-06T19:27:12.0161404Z    [1m[94m= [0m[1mnote[0m: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
2026-07-06T19:27:12.0162752Z    [1m[94m= [0m[1mhelp[0m: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
2026-07-06T19:27:12.0163203Z 
2026-07-06T19:27:12.0163409Z [1m[91merror[0m[1m: unresolved link to `count`[0m
2026-07-06T19:27:12.0163859Z   [1m[94m--> [0mrust_scorer/src/multi_score.rs:62:7
2026-07-06T19:27:12.0164562Z    [1m[94m|[0m
2026-07-06T19:27:12.0165019Z [1m[94m62[0m [1m[94m|[0m /// [`count`] equals `1`.
2026-07-06T19:27:12.0165540Z    [1m[94m|[0m       [1m[91m^^^^^[0m [1m[91mno item named `count` in scope[0m
2026-07-06T19:27:12.0165981Z    [1m[94m|[0m
2026-07-06T19:27:12.0166472Z    [1m[94m= [0m[1mhelp[0m: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
2026-07-06T19:27:12.0166876Z 
2026-07-06T19:27:12.0682107Z [1m[91merror[0m: could not document `rust_scorer`
2026-07-06T19:27:12.0780532Z ##[error]Process completed with exit code 101.

--- a/rust_scorer/Cargo.toml
+++ b/rust_scorer/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rust_scorer"
-version = "1.0.7"
+version = "1.1.0"
 edition = "2024"
 license = "Apache-2.0"
 description = "Native CLI binary for scoring NEAT-AI creatures against training data."

--- a/rust_scorer/Cargo.toml
+++ b/rust_scorer/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rust_scorer"
-version = "1.0.6"
+version = "1.0.7"
 edition = "2024"
 license = "Apache-2.0"
 description = "Native CLI binary for scoring NEAT-AI creatures against training data."

--- a/rust_scorer/src/multi_score.rs
+++ b/rust_scorer/src/multi_score.rs
@@ -46,6 +46,48 @@ use crate::stream_score::{activation_worker_count_for_scorer, effective_fused_re
 /// Keep aligned with main scorer formula.
 const GROWTH_COST: f64 = 0.000_000_1;
 
+/// Instrumentation for the "single pass over training data" invariant
+/// (Issue #3236).
+///
+/// Every multi-creature batch scoring path in this module sweeps the training
+/// corpus **exactly once** via
+/// [`neat_core::training_bin_stream::for_each_read_chunk`], scoring the whole
+/// batch of creatures in that single pass. This process-global counter lets the
+/// `single_pass_assertion` integration test observe how many sweeps a scoring
+/// invocation performs and fail loudly if a future change reintroduces
+/// per-creature re-reads of the training data (which would make the sweep count
+/// scale with creature count instead of staying flat at `1`).
+///
+/// Usage from a test: call [`reset`], run one scoring invocation, then assert
+/// [`count`] equals `1`.
+///
+/// Production overhead is a single atomic increment per whole scoring run —
+/// negligible against a full training-corpus sweep.
+pub mod training_pass_probe {
+    use std::sync::atomic::{AtomicU64, Ordering};
+
+    static TRAINING_DATA_SWEEPS: AtomicU64 = AtomicU64::new(0);
+
+    /// Reset the sweep counter to zero. Call immediately before a scoring
+    /// invocation whose sweep count you want to observe.
+    #[allow(dead_code)] // used by the `single_pass_assertion` integration test.
+    pub fn reset() {
+        TRAINING_DATA_SWEEPS.store(0, Ordering::SeqCst);
+    }
+
+    /// Number of full training-data sweeps recorded since the last [`reset`].
+    #[allow(dead_code)] // used by the `single_pass_assertion` integration test.
+    pub fn count() -> u64 {
+        TRAINING_DATA_SWEEPS.load(Ordering::SeqCst)
+    }
+
+    /// Record one full sweep over the training corpus. Called immediately
+    /// before each `for_each_read_chunk` invocation in the batch paths.
+    pub(crate) fn record_sweep() {
+        TRAINING_DATA_SWEEPS.fetch_add(1, Ordering::SeqCst);
+    }
+}
+
 struct LoadedCreature {
     key: String,
     path: PathBuf,
@@ -393,6 +435,10 @@ pub fn score_from_creature_dir(
         Ok(())
     };
 
+    // Issue #3236: one sweep over the whole training corpus scores every
+    // creature in the batch. Record it so `single_pass_assertion` fails loudly
+    // if a future change reintroduces per-creature re-reads.
+    training_pass_probe::record_sweep();
     for_each_read_chunk(&bin_files, fused_read_buf_len, |chunk| {
         run_io_loop(
             chunk,
@@ -677,6 +723,8 @@ pub fn score_from_creature_dir_gpu(
             Ok(())
         };
 
+        // Issue #3236: single sweep over the corpus for the whole GPU batch.
+        training_pass_probe::record_sweep();
         for_each_read_chunk(&bin_files, fused_read_buf_len, |chunk| {
             run_io_loop(
                 chunk,
@@ -771,6 +819,9 @@ pub fn score_from_creature_dir_gpu(
             Ok(())
         };
 
+        // Issue #3236: single sweep over the corpus for the whole pipelined
+        // GPU batch.
+        training_pass_probe::record_sweep();
         for_each_read_chunk(&bin_files, fused_read_buf_len, |chunk| {
             run_io_loop(
                 chunk,

--- a/rust_scorer/tests/single_pass_assertion.rs
+++ b/rust_scorer/tests/single_pass_assertion.rs
@@ -1,0 +1,132 @@
+//! Issue #3236: permanent automated assertion that multi-creature
+//! (directory/batch) scoring makes **exactly one pass** over the training data,
+//! scoring the whole batch in that single sweep — independent of the number of
+//! creatures.
+//!
+//! The assertion is deterministic: `multi_score` records every sweep over the
+//! training corpus through
+//! [`rust_scorer::multi_score::training_pass_probe`], so these tests reset the
+//! probe, run one scoring invocation, then assert the observed sweep count is
+//! `1`. If a future change reintroduces per-creature re-reads of the training
+//! data (looping the `for_each_read_chunk` sweep once per creature), the sweep
+//! count would equal the creature count `N` and the `assert_eq!` below fails —
+//! breaking `cargo test` before the regression can land.
+
+use std::io::Write;
+use std::path::Path;
+use std::sync::Mutex;
+
+use rust_scorer::cost::CostKind;
+use rust_scorer::gpu::GpuBackendLabel;
+use rust_scorer::multi_score::{score_from_creature_dir, training_pass_probe};
+
+/// The sweep probe is a process-global counter, so tests in this binary that
+/// touch it must not run concurrently. Serialise them through this lock; recover
+/// from poisoning so one failing test does not cascade into the others.
+static PROBE_LOCK: Mutex<()> = Mutex::new(());
+
+/// A minimal forward-only identity creature (1 input, 1 output). Every creature
+/// in a directory-mode batch must share the same input/output shape.
+fn minimal_creature() -> &'static str {
+    r#"{"input":1,"output":1,"forwardOnly":true,"neurons":[{"type":"output","uuid":"output-0","bias":0.0,"squash":"IDENTITY"}],"synapses":[{"fromUUID":"input-0","toUUID":"output-0","weight":1.0}]}"#
+}
+
+/// Write `n` distinct creature JSON files (identical shape, distinct filenames)
+/// into `creatures_dir`.
+fn write_creatures(creatures_dir: &Path, n: usize) {
+    for i in 0..n {
+        std::fs::write(
+            creatures_dir.join(format!("creature-{i}.json")),
+            minimal_creature(),
+        )
+        .expect("write creature json");
+    }
+}
+
+/// Write a training corpus of `records` (input, output) pairs across the given
+/// number of `.bin` files, round-robin, so multi-file traversal is exercised.
+fn write_training_data(data_dir: &Path, records: &[(f32, f32)], bin_files: usize) {
+    let files: Vec<std::fs::File> = (0..bin_files)
+        .map(|i| std::fs::File::create(data_dir.join(format!("{i}.bin"))).expect("create bin"))
+        .collect();
+    let mut files = files;
+    for (idx, (input, output)) in records.iter().enumerate() {
+        let f = &mut files[idx % bin_files];
+        f.write_all(&input.to_le_bytes()).expect("write input");
+        f.write_all(&output.to_le_bytes()).expect("write output");
+    }
+}
+
+/// Score `n_creatures` against a fixed corpus on the CPU path and return the
+/// number of training-data sweeps observed for that single invocation.
+fn sweeps_for_batch(n_creatures: usize, bin_files: usize) -> u64 {
+    let tmp = tempfile::tempdir().expect("create tempdir");
+    let creatures_dir = tmp.path().join("creatures");
+    let data_dir = tmp.path().join("data");
+    std::fs::create_dir(&creatures_dir).expect("create creatures dir");
+    std::fs::create_dir(&data_dir).expect("create data dir");
+
+    write_creatures(&creatures_dir, n_creatures);
+    let records: Vec<(f32, f32)> = (0..64).map(|i| (i as f32 * 0.1, i as f32 * 0.1)).collect();
+    write_training_data(&data_dir, &records, bin_files);
+
+    training_pass_probe::reset();
+    let scores = score_from_creature_dir(
+        &creatures_dir,
+        &data_dir,
+        GpuBackendLabel::CpuFallback,
+        CostKind::Mse,
+    )
+    .expect("scoring must succeed");
+
+    // Sanity: the single sweep really scored every creature in the batch.
+    assert_eq!(
+        scores.len(),
+        n_creatures,
+        "expected a score for each of the {n_creatures} creatures",
+    );
+
+    training_pass_probe::count()
+}
+
+/// Core assertion: scoring N > 1 creatures traverses the training data exactly
+/// once, regardless of N. A regression that re-reads per creature would report
+/// `N` sweeps instead of `1`.
+#[test]
+fn multi_creature_batch_makes_exactly_one_pass() {
+    let _guard = PROBE_LOCK.lock().unwrap_or_else(|e| e.into_inner());
+
+    for &n in &[2_usize, 5, 11] {
+        let sweeps = sweeps_for_batch(n, 1);
+        assert_eq!(
+            sweeps, 1,
+            "scoring {n} creatures must make exactly one pass over the training data, observed {sweeps}",
+        );
+    }
+}
+
+/// The one-pass property is independent of how many `.bin` files the corpus is
+/// split across: a single sweep still covers the whole multi-file corpus.
+#[test]
+fn multi_creature_batch_one_pass_across_multiple_bin_files() {
+    let _guard = PROBE_LOCK.lock().unwrap_or_else(|e| e.into_inner());
+
+    let sweeps = sweeps_for_batch(4, 3);
+    assert_eq!(
+        sweeps, 1,
+        "one sweep must cover the whole multi-file corpus, observed {sweeps}",
+    );
+}
+
+/// Contrast case (Scope item 3): the single-creature path also makes exactly
+/// one pass, pinning that N == 1 and N > 1 share the same one-sweep contract.
+#[test]
+fn single_creature_makes_exactly_one_pass() {
+    let _guard = PROBE_LOCK.lock().unwrap_or_else(|e| e.into_inner());
+
+    let sweeps = sweeps_for_batch(1, 1);
+    assert_eq!(
+        sweeps, 1,
+        "scoring a single creature must make exactly one pass, observed {sweeps}",
+    );
+}

